packages/hardhat/contracts/CityDaoParcel.sol

```diff
@@ -35,30 +38,62 @@ contract CityDaoParcel is ERC721, Ownable {
       public
       returns (uint256)
   {
-      require(msg.sender == _toAddress);
+      if(msg.sender != _toAddress) revert("You must purchase the parcel for yourself!");
+      if(isSold(parcelId)) revert("This parcel has already been sold!");
       
       _safeMint(_toAddress, parcelId);
       _setTokenURI(parcelId, _parcelIdToTokenURI[parcelId]);
 
       delete _parcelIdToPrice[parcelId];
-      delete _parcelIdToTokenURI[parcelId];
+      _parcelIdToSoldStatus[parcelId] = true;
 
       return parcelId;
   }
 
-  function getListedParcels() public view returns (string[] memory) {
-    string[] memory ret = new string[](_parcelIds.length);
-    for (uint i = 0; i < _parcelIds.length; i++) {
-        ret[i] = _parcelIdToTokenURI[i];
-    }
-    return ret;
+  function isSold(uint256 parcelId) public view returns (bool) {
+    return _parcelIdToSoldStatus[parcelId];
   }
-
   function getPrice(uint256 parcelId) public view returns (string memory) {
     return _parcelIdToPrice[parcelId];
   }
 
-  function getParcelIds() public view returns (uint256[] memory) {
-    return _parcelIds;
+  function getListedParcelURIs() public view returns (string[] memory) {
+    uint256[] memory listedParcelIds = getListedParcelIds();
+    string[] memory ret = new string[](listedParcelIds.length);
+    for (uint i = 0; i < listedParcelIds.length; i++) {
+      if (_parcelIdToSoldStatus[listedParcelIds[i]] == false) {
+        ret[i] = _parcelIdToTokenURI[listedParcelIds[i]];
+      }
+    }
+    return ret;
+  }
+  function getSoldParcelURIs() public view returns (string[] memory) {
+    uint256[] memory soldParcelIds = getSoldParcelIds();
+    string[] memory ret = new string[](soldParcelIds.length);
+    for (uint i = 0; i < soldParcelIds.length; i++) {
+      if (_parcelIdToSoldStatus[soldParcelIds[i]] == true) {
+        ret[i] = _parcelIdToTokenURI[soldParcelIds[i]];
+      }
+    }
+    return ret;
+  }
+
+  function getListedParcelIds() public view returns (uint256[] memory) {
+    uint256[] memory ret = new uint256[](_parcelIds.length);
+    for (uint i = 0; i < _parcelIds.length; i++) {
+      if (_parcelIdToSoldStatus[_parcelIds[i]] == false) {
+        ret[i] = _parcelIds[i];
+      }
+    }
+    return ret;
+  }
+  function getSoldParcelIds() public view returns (uint256[] memory) {
+    uint256[] memory ret = new uint256[](_parcelIds.length);
+    for (uint i = 0; i < _parcelIds.length; i++) {
+      if (_parcelIdToSoldStatus[_parcelIds[i]] == true) {
+        ret[i] = _parcelIds[i];
```

Comment:
it might be better practice to avoid the for loops and make the frontend do more work - example to get sold parcel IDs

getSoldParcels -> just returns array, no loop
isParcelSold() -> frontend calls this for each

but we should check with solidity pro

Reply:
Actually, I agree that's much simpler. In favor of changing it